### PR TITLE
feat: add usePaymentMethods hook (ACC-6917)

### PR DESCRIPTION
**Pull request by OnurVar**
## Summary

- Add `usePaymentMethods()` hook that merges payment method availability data with display resources (logos, colors, names) into a unified `PaymentMethodItem` type
- Extend `PrimerCheckoutContextValue` with `paymentMethodResources` and `isLoadingResources`
- Extend `PrimerCheckoutProvider` to eagerly fetch payment method resources via `AssetsManager` after SDK init
- Support filtering (include/exclude), sorting (showCardFirst), selection state, and optional surcharge data
- Export `usePaymentMethods`, `PaymentMethodItem`, `UsePaymentMethodsOptions`, `UsePaymentMethodsReturn` from package root

## Jira

[ACC-6917](https://primerapi.atlassian.net/browse/ACC-6917)

## Test plan

- [x] `npm run lint` — clean
- [x] `npm test` — 32/32 tests pass (11 new for usePaymentMethods)
- [ ] Manual verification: hook returns merged payment methods with logos, names, colors inside PrimerCheckoutProvider

[ACC-6917]: https://primerapi.atlassian.net/browse/ACC-6917

### CODE REVIEW

File: src/Components/types/PrimerCheckoutProviderTypes.ts [hunk 1]
```diff
@@ -31,4 +35,6 @@ export interface PrimerCheckoutContextValue {
   error: PrimerError | null;
   clientSession: PrimerClientSession | null;
   availablePaymentMethods: IPrimerHeadlessUniversalCheckoutPaymentMethod[];
```
Comment (bthebladeprimer) [inline src/Components/types/PrimerCheckoutProviderTypes.ts:37]: nit: `IPrimerPaymentMethodResource[]` is too loose here. The native API returns the discriminated union, and the hook already relies on that by narrowing via `'paymentMethodLogo' in resource`. Consider typing this as:

```ts
paymentMethodResources: Array<PrimerPaymentMethodAsset | PrimerPaymentMethodNativeView>;
```

That removes the `as PrimerPaymentMethodAsset` cast in `usePaymentMethods.ts` and lets TS narrow correctly on the discriminant.

File: src/Components/hooks/usePaymentMethods.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    if (!hasFiredLoadRef.current) {
+      hasFiredLoadRef.current = true;
+      onLoadRef.current?.(paymentMethods);
+    }
+  }, [isLoading, paymentMethods]);
+
+  const selectMethod = useCallback((method: PaymentMethodItem | null) => {
+    setSelectedMethod(method);
+  }, []);
+
```
Comment (bthebladeprimer) [inline src/Components/hooks/usePaymentMethods.ts:158]: `selectedMethod` stores a `PaymentMethodItem` reference, but that array is rebuilt inside `useMemo` whenever resources arrive, filters change, or the client session updates. After a rebuild the stored reference is orphaned — it no longer points to an item in the current `paymentMethods` list, so equality checks like `paymentMethods.includes(selectedMethod)` or `item === selectedMethod` will silently be false.

Two options:
1. Track the selection by `type` (string) and derive `selectedMethod` from the current list: `paymentMethods.find(m => m.type === selectedType) ?? null`.
2. Keep it as-is but reconcile in an effect: if the selected type is no longer present (e.g. filtered out), clear it.

File: src/Components/PrimerCheckoutProvider.tsx [hunk 3]
```diff
@@ -146,8 +148,25 @@ export function PrimerCheckoutProvider({
           ...prev,
           isReady: true,
           availablePaymentMethods: methods,
+          isLoadingResources: true,
```
Comment (bthebladeprimer) [inline src/Components/PrimerCheckoutProvider.tsx:163]: When the provider re-initializes (new `clientToken`), `isLoadingResources` flips back to `true` but the previous `paymentMethodResources` array is kept until the new fetch resolves. `isLoading` in the hook gates output correctly, but it's worth clearing the array here so we don't have a transient mismatch between the fresh `availablePaymentMethods` and stale resources:

```ts
setState((prev) => ({
  ...prev,
  isReady: true,
  availablePaymentMethods: methods,
  paymentMethodResources: [],
  isLoadingResources: true,
}));
```

File: src/Components/PrimerCheckoutProvider.tsx [hunk 4]
```diff
@@ -146,8 +148,25 @@ export function PrimerCheckoutProvider({
...
+        if (!cancelled) {
+          setState((prev) => ({
+            ...prev,
+            paymentMethodResources: resources,
+            isLoadingResources: false,
+          }));
+        }
+      } catch (err) {
+        console.warn('[Primer] Failed to load payment method resources', err);
+        if (!cancelled) {
```
Comment (bthebladeprimer) [inline src/Components/PrimerCheckoutProvider.tsx]: The resource-fetch failure is `console.warn`'d and swallowed — consumers can't tell assets failed to load. After the catch, `isLoadingResources` goes false with empty resources, so the hook renders methods with fallback names and no logos, indistinguishable from a success case with no assets.

If that's intentional (don't block checkout), fine — but consider surfacing it somewhere, e.g. a `resourcesError` field on the context, so UIs can show a retry affordance.

File: src/Components/PrimerCheckoutProvider.tsx [hunk 5]
```diff
@@ -4,17 +4,19 @@ import { ThemeContext } from './internal/theme/ThemeContext';
...
 import { PrimerHeadlessUniversalCheckout } from '../HeadlessUniversalCheckout/PrimerHeadlessUniversalCheckout';
+import PrimerHeadlessUniversalCheckoutAssetsManager from '../HeadlessUniversalCheckout/Managers/AssetsManager';
 import type { PrimerSettings } from '../models/PrimerSettings';
-import type {
-  PrimerCheckoutProviderProps,
-  PrimerCheckoutContextValue,
-} from '../models/components/PrimerCheckoutProviderTypes';
+import type { PrimerCheckoutProviderProps, PrimerCheckoutContextValue } from './types/PrimerCheckoutProviderTypes';
+
+const assetsManager = new PrimerHeadlessUniversalCheckoutAssetsManager();
```
Comment (bthebladeprimer) [inline src/Components/PrimerCheckoutProvider.tsx:12]: nit: module-level singleton — shared across all `PrimerCheckoutProvider` instances and instantiated at import time. Probably fine in practice, but a stateful native manager at module scope can complicate teardown/tests. Worth considering instantiating inside the provider or lazily.

File: src/Components/types/PrimerCheckoutProviderTypes.ts (hunk 1 above)
Comment (OnurVar) [reply src/Components/types/PrimerCheckoutProviderTypes.ts:37]: Done — typed `paymentMethodResources` as `Array<PrimerPaymentMethodAsset | PrimerPaymentMethodNativeView>` on the context and `PaymentMethodItem.resource`, and updated `AssetsManager.getPaymentMethodResources` to return the union. Cast in the hook is gone; TS narrows on the discriminant.

File: src/Components/hooks/usePaymentMethods.ts (hunk 2 above)
Comment (OnurVar) [reply src/Components/hooks/usePaymentMethods.ts:158]: Done — went with option 1. Now storing `selectedType: string` and deriving `selectedMethod` via `paymentMethods.find(m => m.type === selectedType) ?? null`. Selection survives memo rebuilds and clears automatically if the type is filtered out. Added two tests covering both behaviours.

File: src/Components/PrimerCheckoutProvider.tsx (hunk 3 above)
Comment (OnurVar) [reply src/Components/PrimerCheckoutProvider.tsx:163]: Done — added `paymentMethodResources: []` to the setState when flipping `isLoadingResources: true`. State is now consistent: `isLoadingResources === true` ⇒ empty resources array, never stale.

File: src/Components/PrimerCheckoutProvider.tsx (hunk 4 above)
Comment (OnurVar) [reply src/Components/PrimerCheckoutProvider.tsx]: Done — added `resourcesError: Error | null` to the context and surfaced it through `usePaymentMethods` alongside the existing `error`. The catch now populates it instead of silently `console.warn`ing, and it clears on a fresh fetch. Kept it separate from `error` since that one is for payment-flow failures (`PrimerError`) and has different UX implications.

File: src/Components/PrimerCheckoutProvider.tsx (hunk 5 above)
Comment (OnurVar) [reply src/Components/PrimerCheckoutProvider.tsx:12]: Done — moved `assetsManager` inside the provider via `useState(() => new PrimerHeadlessUniversalCheckoutAssetsManager())` so each provider instance owns its own, created lazily and cleaned up with the component.

File: src/Components/hooks/usePaymentMethods.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+      // Filter: include first, then exclude
+      if (include && include.length > 0) {
+        result = result.filter((item) => include.includes(item.type));
+      }
+      if (exclude && exclude.length > 0) {
+        result = result.filter((item) => !exclude.includes(item.type));
+      }
+
+      // Sort: PAYMENT_CARD first
+      if (showCardFirst) {
```
Comment (semirp) [inline src/Components/hooks/usePaymentMethods.ts]: [q]: was there ever a need or request for this?

File: src/Components/types/PaymentMethodTypes.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import type { PrimerPaymentMethodAsset, PrimerPaymentMethodNativeView } from '../../models/PrimerPaymentMethodResource';
+import type { IPrimerHeadlessUniversalCheckoutPaymentMethod } from '../../models/PrimerHeadlessUniversalCheckoutPaymentMethod';
+
+export interface PaymentMethodItem {
+  type: string;
+  name: string;
+  logo?: string;
+  backgroundColor?: string;
+  isNativeView: boolean;
```
Comment (semirp) [inline src/Components/types/PaymentMethodTypes.ts]: [q]: can we discriminate by `resource` already?

File: src/models/PrimerClientSession.ts [hunk 8]
```diff
@@ -8,6 +8,19 @@ interface IPrimerClientSession {
...
+
+export type PrimerClientSessionPaymentMethodOption = IPrimerClientSessionPaymentMethodOption;
+
+interface IPrimerClientSessionPaymentMethodOption {
+  surcharge?: IPrimerClientSessionSurcharge;
+}
+
+export type PrimerClientSessionSurcharge = IPrimerClientSessionSurcharge;
+
+interface IPrimerClientSessionSurcharge {
```
Comment (semirp) [inline src/models/PrimerClientSession.ts:31]: surcharges can be per payment method (e.g. PAYPAL, APPLE_PAY...) or per card network in case it's set for the `PAYMENT_CARD`...

File: src/Components/types/PaymentMethodTypes.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  nativeViewName?: string;
+  categories: string[];
```
Comment (nquinn-primer) [inline src/Components/types/PaymentMethodTypes.ts]: categories and intents are both Unions (`'NATIVE_UI' | 'RAW_DATA' | 'CARD_COMPONENTS'` and `'CHECKOUT' | 'VAULT'`) which we can use here?

File: src/Components/hooks/usePaymentMethods.ts [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+      }
+
+      let result: PaymentMethodItem[] = availablePaymentMethods.map((method) => {
+        const resource = resourceMap.get(method.paymentMethodType);
+
+        let logo: string | undefined;
+        let backgroundColor: string | undefined;
+
+        if (resource && 'paymentMethodLogo' in resource) {
+          logo = resource.paymentMethodLogo.colored ?? resource.paymentMethodLogo.light;
```
Comment (nquinn-primer) [inline src/Components/hooks/usePaymentMethods.ts]: We never use the dark resource, intended?

File: src/Components/hooks/usePaymentMethods.ts (hunk 6 above)
Comment (OnurVar) [reply src/Components/hooks/usePaymentMethods.ts]: Removed. No real use case, and consumers can sort themselves if they want card first. Default is now API order.

File: src/Components/types/PaymentMethodTypes.ts (hunk 7 above)
Comment (OnurVar) [reply src/Components/types/PaymentMethodTypes.ts]: Yes — removed `isNativeView`. `resource` is the discriminator now:

```ts
if (item.resource && 'nativeViewName' in item.resource) { /* native view */ }
if (item.resource && 'paymentMethodLogo' in item.resource) { /* asset */ }
```

Matches how Android's `PrimerPaymentMethodResource` sealed interface works.

File: src/models/PrimerClientSession.ts (hunk 8 above)
Comment (OnurVar) [reply src/models/PrimerClientSession.ts:31]: Good catch — modeled both. `IPrimerClientSessionPaymentMethodOption` now has `networks` for PAYMENT_CARD. Exposed on `PaymentMethodItem.surcharge` as a discriminated union mirroring Android's `Surcharge` sealed interface:

```ts
type PaymentMethodSurcharge =
  | { kind: 'flat'; amount: number }           // PAYPAL, APPLE_PAY, …
  | { kind: 'perNetwork'; amounts: Record<string, number> };  // PAYMENT_CARD
```

Tests cover both paths + the edge cases (networks on non-card ignored, top-level surcharge on card ignored in favor of networks).

File: src/Components/types/PaymentMethodTypes.ts (hunk 9 above)
Comment (OnurVar) [reply src/Components/types/PaymentMethodTypes.ts]: Done — typed them as `PrimerPaymentMethodManagerCategoryName[]` and `PrimerSessionIntentName[]` (the existing unions on `IPrimerHeadlessUniversalCheckoutPaymentMethod`). Consumers now get autocomplete and typo protection.

File: src/Components/hooks/usePaymentMethods.ts (hunk 10 above)
Comment (OnurVar) [reply src/Components/hooks/usePaymentMethods.ts]: Good catch — it wasn't. Now theme-aware via `useColorScheme`:

```ts
const themeVariant = isDark ? asset.dark : asset.light;
logo = asset.colored ?? themeVariant ?? asset.light ?? asset.dark;
```

Same logic applied to `backgroundColor`. `colored` still wins when present; otherwise we pick the variant matching the color scheme and cross-fall-back so a missing theme variant doesn't leave the asset undefined. Consumers wanting full control can still read `item.resource.paymentMethodLogo` (all three variants exposed on `IPrimerAsset`).

**Review by bthebladeprimer — APPROVED** [review]

File: src/Components/types/PaymentMethodTypes.ts (hunk 9 above)
Comment (semirp) [reply src/Components/types/PaymentMethodTypes.ts]: `CARD_COMPONENTS` is not used anymore right? we have removed that integration 2 years ago

**Review by nquinn-primer — APPROVED** [review]
